app/src/full/java/io/homeassistant/companion/android/sensors/ActivitySensorManager.kt

```diff
@@ -0,0 +1,197 @@
+import io.homeassistant.companion.android.domain.integration.SensorRegistration
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class ActivitySensorManager : BroadcastReceiver(), SensorManager {
+
+    companion object {
+
+        internal const val TAG = "ActivitySM"
+
+        const val ACTION_REQUEST_ACTIVITY_UPDATES =
+            "io.homeassistant.companion.android.background.REQUEST_ACTIVITY_UPDATES"
+        const val ACTION_UPDATE_ACTIVITY =
+            "io.homeassistant.companion.android.background.UPDATE_ACTIVITY"
+
+        fun restartActivityTracking(context: Context) {
+            val intent = Intent(context, ActivitySensorManager::class.java)
+            intent.action = ACTION_REQUEST_ACTIVITY_UPDATES
+
+            context.sendBroadcast(intent)
+        }
+
+        private val activity = SensorManager.BasicSensor(
+            "detected_activity",
+            "sensor",
+            "Detected Activity"
+        )
+
+        private var stored_activity: String = "unknown"
+        private var stored_attributes = mutableMapOf<String, Int>()
+    }
+
+    @Inject
+    lateinit var integrationUseCase: IntegrationUseCase
+
+    private val ioScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+
+    override fun onReceive(context: Context, intent: Intent) {
+        ensureInjected(context)
+
+        when (intent.action) {
+            Intent.ACTION_BOOT_COMPLETED,
+            ACTION_REQUEST_ACTIVITY_UPDATES -> setupActivityTracking(context)
+            ACTION_UPDATE_ACTIVITY -> handleActivityUpdate(intent, context)
+            else -> Log.w(TAG, "Unknown intent action: ${intent.action}!")
+        }
+    }
+
+    private fun ensureInjected(context: Context) {
+        if (context.applicationContext is GraphComponentAccessor) {
+            DaggerSensorComponent.builder()
+                .appComponent((context.applicationContext as GraphComponentAccessor).appComponent)
+                .build()
+                .inject(this)
+        } else {
+            throw Exception("Application Context passed is not of our application!")
+        }
+    }
+
+    private fun setupActivityTracking(context: Context) {
+        if (!checkPermission(context)) {
+            Log.w(TAG, "Not starting activity tracking because of permissions.")
+            return
+        }
+
+        val sensorDao = AppDatabase.getInstance(context).sensorDao()
+
+        ioScope.launch {
+            try {
+                Log.d(TAG, "Unregistering for activity updates.")
+                ActivityRecognition.getClient(context).removeActivityUpdates(getPendingIntent(context))
+
+                if (sensorDao.get(activity.id)?.enabled == true) {
+                    Log.d(TAG, "Registering for activity updates.")
+
+                    ActivityRecognition.getClient(context).requestActivityUpdates(12000, getPendingIntent(context))
```

Comment:
Isn't this requesting every 12 seconds?

Reply:
Do we instead want to change to use `requestActivityTransitionUpdates` since we really only care about changes activity type?  This should get updated less frequently and cause less of a drain but provide the same basic support.

Reply:
It is, missed one more 0. 🤦 
According to the api documentation, `requestActivityUpdates` is more power efficient cause it does not track transitional conditions (enter, exit)

> The activities are detected by periodically waking up the device and reading short bursts of sensor data. It only makes use of low power sensors in order to keep the power usage to a minimum. 
>The activity detection update interval can be controlled with the detectionIntervalMillis parameter. Larger values will result in fewer activity detections while improving battery life. Smaller values will result in more frequent activity detections but will consume more power since the device must be woken up more frequently.

Reply:
I'am still thinking of moving off restartActivityTracking() fun from SensorReceiver.OnReceive. It may trigger the next activity update even before the UpdateSensors function finished. I tried to call restartActivityTracking() in Application class with interval 60 sec and it worked ok. 
@JBassett could you recommend a better place where to call it to correspond sensor's enabled\disabled state?